### PR TITLE
SEO-351 Refactored RobotsTxt extension

### PR DESCRIPTION
**Pull request by gabrys**
- The `wikia-robots-txt.php` logic is now moved to `WikiaRobots`.
- `RobotsTxt` is now renamed to `RobotsTxtBuilder` and reads no context from outside
- `PathBuilder` was extracted with responsibility to build paths for `RobotsTxtBuilder`
- `PathBuilder` reads `$wgContLang` and `$wgArticlePath` to build the paths
- `WikiaRobots` reads all sorts of wg vars to control what's blocked and what's allowed
- Unit tests were added to all three components
- Ability to substitute the `robots.txt` file for a specific wiki was removed
- Experimental method `WikiaRobots::getMetaRobotsPolicy` added, ready for SEO-113

This refactoring is a step towards having the `robots.txt` and `<meta name="robots">` controlled through one class: `WikiaRobots`.

The class builds its state in its constructor, computing the blocked namespaces, blocked params and allowed special pages.

It then uses its state to configure a `RobotsBuilder` instance used to generate the `robots.txt` file, but also in the `getMetaRobotsPolicy` method that can be used to query for the expected `<meta>` tag. This means there's one place to configure the behavior robots should follow on Wikia pages.

### CODE REVIEW

File: extensions/wikia/RobotsTxt/WikiaRobots.class.php [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+
+		// Block namespaces
+		foreach ( $this->blockedNamespaces as $ns ) {
+			$robots->addDisallowedPaths(
+				$this->pathBuilder->buildPathsForNamespace( $ns )
+			);
+		}
+
+		// Block user pages
+		$robots->addDisallowedPaths( [ '/d/u/' ] );
```
Comment (saipetch) [inline extensions/wikia/RobotsTxt/WikiaRobots.class.php]: consider extracting to member?

File: extensions/wikia/RobotsTxt/WikiaRobots.class.php [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+
+		// SEO-302: Allow Googlebot to crawl Android app contents
+		// @see http://developer.android.com/training/app-indexing/enabling-app-indexing.html)
+		// The order of precedence between those two is undefined:
+		// "Disallow: /*?*action=" and "Allow: /api.php"
+		// @see https://developers.google.com/webmasters/control-crawl-index/docs/robots_txt#order-of-precedence-for-group-member-records
+		// That's why we're adding quite explicit "Allow: /api.php?*action=" (even though it's redundant)
+		// robots.txt Tester in Google Search Console shows this will do:
+		// @see https://www.google.com/webmasters/tools/robots-testing-tool?hl=en&siteUrl=http://muppet.wikia.com/
+		$robots->addAllowedPaths( [ '/api.php?', '/api.php?action=', '/api.php?*&action=' ] );
```
Comment (saipetch) [inline extensions/wikia/RobotsTxt/WikiaRobots.class.php]: consider extracting to member?

File: extensions/wikia/RobotsTxt/WikiaRobots.class.php [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+
+		// Block robots
+		$robots->addBlockedRobots( $this->blockedRobots );
+
+		// Deprecated items, probably we should delete them
+		$robots->addDisallowedPaths( [
```
Comment (saipetch) [inline extensions/wikia/RobotsTxt/WikiaRobots.class.php]: consider extracting to member?

File: extensions/wikia/RobotsTxt/WikiaRobots.class.php [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+		if ( isset( $wgRobotsTxtCustomRules['disallowNamespace'] ) ) {
+			$this->blockedNamespaces = array_merge(
+				$this->blockedNamespaces,
+				(array) $wgRobotsTxtCustomRules['disallowNamespace']
+			);
+		}
+	}
+
+	public function configureRobotsBuilder( RobotsTxtBuilder $robots ) {
+		global $wgEnableSpecialSitemapExt,
```
Comment (saipetch) [inline extensions/wikia/RobotsTxt/WikiaRobots.class.php]: consider merging to one line?

File: extensions/wikia/RobotsTxt/PathBuilder.class.php [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+	 * Get the special page's aliases
+	 *
+	 * @param string $pageName special page name as specified in the SpecialPage object
+	 * @return array
+	 */
+	private function getSpecialPageNames( $pageName ) {
+		$aliases = [];
+		foreach ( $this->languages as $lang ) {
+			$specialAliases = $lang->getSpecialPageAliases();
+			if ( isset( $specialAliases[$pageName] ) ) {
```
Comment (saipetch) [inline extensions/wikia/RobotsTxt/PathBuilder.class.php]: Do we have empty value for special alias?

File: extensions/wikia/RobotsTxt/RobotsTxtBuilder.class.php [hunk 6]
```diff
@@ -0,0 +1,126 @@
+<?php
+
+class RobotsTxtBuilder {
+
+	private $allowed = [];
+	private $blockedRobots = [];
+	private $disallowed = [];
+	private $sitemap;
+	private $experimentalAllowDisallowSection = null;
```
Comment (saipetch) [inline extensions/wikia/RobotsTxt/RobotsTxtBuilder.class.php]: Clean up?

File: extensions/wikia/RobotsTxt/tests/RobotsTxtBuilderTest.php [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+				]
+			],
+		];
+	}
+
+	/**
+	 * Test the methods called more times
+	 *
+	 * @covers       addAllowedPaths::addAllowPaths
+	 * @dataProvider dataProviderClassApi
```
Comment (saipetch) [inline extensions/wikia/RobotsTxt/tests/RobotsTxtBuilderTest.php]: remove @dataProvider ?

File: extensions/wikia/RobotsTxt/tests/WikiaRobotsTest.php [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+			->willReturnCallback( function ( $param ) {
+				return [ '/path/for/param/' . $param ];
+			} );
+		return $pathBuilderMock;
+	}
+
+	/**
+	 * Test WikiaRobots builds a "Disallow: /" robots.txt on dev environment
+	 */
+	public function testDevEnvironment() {
```
Comment (saipetch) [inline extensions/wikia/RobotsTxt/tests/WikiaRobotsTest.php]: staging?

File: extensions/wikia/RobotsTxt/PathBuilder.class.php (hunk 5 above)
Comment (gabrys) [inline extensions/wikia/RobotsTxt/PathBuilder.class.php]: There's a hook that allows 3rd party extensions to break the structure, I'm adding an additional is_array here to be safe.

File: extensions/wikia/RobotsTxt/WikiaRobots.class.php (hunk 3 above)
Comment (gabrys) [inline extensions/wikia/RobotsTxt/WikiaRobots.class.php]: The plan is to remove it altogether, so I'm gonna leave it for now

**Comment by saipetch** [comment]
LGTM

**Pull request merged**